src/porepy/examples/geothermal_reservoir.py

```diff
@@ -0,0 +1,493 @@
+import logging
+from typing import TYPE_CHECKING, Callable, cast
+
+import numpy as np
+from numpy.typing import NDArray
+
+import porepy as pp
+from porepy.applications.boundary_conditions.model_boundary_conditions import (
+    BoundaryConditionsMechanicsNeumann,
+    HydrostaticBoundaryPressureValues,
+    LithostaticBoundaryStressValues,
+    ThermalGradientBoundaryTemperatureValues,
+)
+from porepy.applications.initial_conditions.model_initial_conditions import (
+    InitialConditionHydrostaticPressureValues,
+    InitialConditionThermalGradientTemperatureValues,
+)
+from porepy.applications.md_grids.model_geometries import (
+    TwoEllipticFractures3d,
+    TwoWells3d,
+)
+from porepy.numerics.nonlinear import line_search
+
+logger = logging.getLogger(__name__)
+
+logging.basicConfig(level=logging.INFO)
+
+
+class WellBoundaryConditions(pp.PorePyModel):
+    """Class defining boundary conditions for geothermal reservoir models
+    with wells and fractures.
+
+    We impose the well protocols as boundary conditions on the appropriate grids. For
+    the boundaries of the 3d matrix grid (and any fracture grids), we impose hydrostatic
+    pressure and a thermal gradient temperature, corresponding to Dirichlet boundary
+    conditions for pressure and temperature. We default Dirichlet conditions for all
+    fluxes and all boundaries.
+
+    Super calls in methods `bc_values_pressure` and `bc_values_temperature` suggest that
+    this class should be used as a mixin with higher MRO priority than classes defining
+    non-well boundary conditions, e.g. `HydrostaticBoundaryPressureValues` and
+    `ThermalGradientBoundaryTemperatureValues`.
+    """
+
+    if TYPE_CHECKING:
+        well_names: list[str]
+        """List of well tags, e.g. ['injection_well', 'production_well']."""
+
+    def bc_values_pressure(self, bg: pp.BoundaryGrid) -> np.ndarray:
+        """Return boundary values for pressure on all boundaries.
+
+        Parameters:
+            sd: The subdomain for which to return the BC values.
+
+        Returns:
+            The boundary values for pressure on the given subdomain.
+        """
+        sd = bg.parent
+        values = super().bc_values_pressure(bg)  # type: ignore[misc]
+        if self.is_well_grid(sd):
+            well_tag = self.well_names[sd.tags["parent_well_index"]]
+            protocol = self.well_protocols()[well_tag]
+            # Find indices of the well boundary sides.
+            domain_sides = self.domain_boundary_sides(bg)
+            inds = domain_sides.top if self.nd == 3 else domain_sides.north
```

Comment:
```suggestion
            # The top of the domain is '.top' in 3d, '.north' in 2d.
            inds = domain_sides.top if self.nd == 3 else domain_sides.north
```
Or something like that - I was puzzled for a little while.